File: backend/services/scanner.py
import os
import time
import threading
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor, as_completed
from sqlalchemy.orm import Session
from backend.core.db import SessionLocal
from backend.models.media import Library, Movie, MovieFile, TVShow, Season, Episode
from backend.services.parser import parse_filename
from backend.services.mediainfo import extract_media_info
from backend.services.nfo_reader import NFOReader
from backend.core.progress import active_scans
from backend.core.task_manager import update_task

SUPPORTED_EXTS = {
    # Common
    '.mkv', '.mp4', '.avi', '.mov', '.wmv', '.m4v', '.ts', '.m2ts', '.mts',
    # Legacy & DVD
    '.mpg', '.mpeg', '.m1v', '.m2v', '.vob', '.divx', '.xvid',
    # Web & Other
    '.webm', '.flv', '.f4v', '.ogv', '.ogm', '.3gp', '.3g2',
    # Disc Images
    '.iso'
}

class ScannerService:
    def __init__(self, db: Session):
        self.db = db
        self.db_lock = threading.Lock()
        self.max_workers = 4  # Safe default for mixed HDD/SSD/NAS environments

    def scan_library(self, library_id: int, task_id: str = None):
        lib = self.db.query(Library).filter(Library.id == library_id).first()
        if not lib:
            active_scans[library_id] = {"status": "error"}
            return {"status": "error", "message": "Library not found"}
            
        normalized_path = lib.path.replace("\\", "/")
        print(f"[Scanner] Starting multi-threaded scan of library {lib.id} at: {normalized_path}")
        
        if not os.path.exists(normalized_path):
            active_scans[library_id] = {"status": "error", "message": f"Path not found: {normalized_path}"}
            if task_id: update_task(task_id, status="error", message=f"Path not found: {normalized_path}")
            return {"status": "error", "message": f"Path not found: {normalized_path}"}
            
        # 1. Pre-fetch Caches
        existing_movies_paths = {}
        existing_episodes_paths = {}
        title_to_movie = {}
        title_to_show = {}
        
        if lib.type == 'movie' or lib.type is None:
            m_files = self.db.query(MovieFile.file_path, MovieFile.movie_id).join(Movie).filter(Movie.library_id == library_id).all()
            existing_movies_paths = {f.file_path: f.movie_id for f in m_files}
            movies = self.db.query(Movie).filter(Movie.library_id == library_id).all()
            title_to_movie = {m.title.lower(): m for m in movies}
            
        if lib.type == 'tv' or lib.type is None:
            e_files = self.db.query(Episode.file_path, Season.show_id).join(Season).join(TVShow).filter(TVShow.library_id == library_id).all()
            existing_episodes_paths = {f.file_path: f.show_id for f in e_files}
            shows = self.db.query(TVShow).filter(TVShow.library_id == library_id).all()
            title_to_show = {s.title.lower(): s for s in shows}

        # 2. Discovery Phase
        files_to_process = []
        SKIP_DIRS = {'.git', 'node_modules', '.actors', '@eaDir', '#recycle'}
        
        if task_id: update_task(task_id, message="Walking folders...")
        
        processed_movie_ids = set()
        processed_show_ids = set()

        for root, dirs, files in os.walk(normalized_path):
            dirs[:] = [d for d in dirs if not d.startswith('.') and d not in SKIP_DIRS]
            for file in files:
                ext = Path(file).suffix.lower()
                if ext in SUPPORTED_EXTS:
                    file_path = os.path.join(root, file).replace("\\", "/")
                    if file_path in existing_movies_paths:
                        processed_movie_ids.add(existing_movies_paths[file_path])
                    elif file_path in existing_episodes_paths:
                        processed_show_ids.add(existing_episodes_paths[file_path])
                    else:
                        files_to_process.append((file_path, file))

        total_new = len(files_to_process)
        if total_new == 0:
            print("[Scanner] No new files found.")
            if task_id: update_task(task_id, status="completed", message="Scan complete (no new files)")
            return {"status": "success", "files_processed": 0, "movie_ids": list(processed_movie_ids), "show_ids": list(processed_show_ids)}

        active_scans[library_id] = {"total": total_new, "current": 0, "status": "processing"}
        print(f"[Scanner] Found {total_new} new files. Gathering metadata in parallel...")

        # 3. Phase 1: Gather Metadata (Parallel)
        gathered_data = []
        with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
            futures = [executor.submit(self._gather_metadata_worker, f_path, f_name) for f_path, f_name in files_to_process]
            for future in as_completed(futures):
                gathered_data.append(future.result())
                if len(gathered_data) % 50 == 0:
                    print(f"[Scanner] Gathered {len(gathered_data)}/{total_new}...")

        # 4. Phase 2: Batch Registration (Sequential & Atomic)
        print(f"[Scanner] Registering {total_new} items to database in batches of 50...")
        for i, item in enumerate(gathered_data):
            f_path, f_name, parsed, size_bytes = item
            
            if lib.type == 'tv' or (lib.type != 'movie' and parsed.is_tv):
                sid = self._register_tv_sequential(lib, f_path, f_name, parsed, size_bytes, title_to_show)
                processed_show_ids.add(sid)
            else:
                mid = self._register_movie_sequential(lib, f_path, f_name, parsed, size_bytes, title_to_movie)
                processed_movie_ids.add(mid)
            
            # Commit in smaller batches (e.g. 50 items) to "breathe" the DB
            # and let the Task Manager update the persistent status table.
            if (i + 1) % 50 == 0 or (i + 1) == total_new:
                active_scans[library_id]["current"] = i + 1
                if task_id:
                    # Pass self.db to update_task so it uses the same transaction
                    update_task(task_id, progress=i + 1, total=total_new, 
                                message=f"Registering: {i + 1}/{total_new}...", db=self.db)
                else:
                    self.db.commit()
                print(f"[Scanner] Registered {i + 1}/{total_new}...")

        # Final commit happens inside the last batch check
        print("[Scanner] Registration phase complete.")

        active_scans[library_id]["status"] = "done"
        if task_id: 
            update_task(task_id, status="completed", message="Scan complete", db=self.db)
        
        return {
            "status": "success", 
            "files_processed": total_new,
            "movie_ids": list(processed_movie_ids),
            "show_ids": list(processed_show_ids)
        }

    def _gather_metadata_worker(self, file_path: str, filename: str):
        parsed = parse_filename(filename)
        size_bytes = os.path.getsize(file_path) if os.path.exists(file_path) else 0
        return (file_path, filename, parsed, size_bytes)

    def _register_movie_sequential(self, lib: Library, file_path: str, filename: str, parsed, size_bytes: int, cache: dict):
        clean_title = parsed.title.strip()
        lower_title = clean_title.lower()
        
        if lower_title in cache:
            movie = cache[lower_title]
        else:
            movie = Movie(library_id=lib.id, title=clean_title, year=parsed.year, status="unmatched")
            self.db.add(movie)
            # Flush assigns an ID without committing to disk (very fast)
            self.db.flush()
            cache[lower_title] = movie

        mfile = MovieFile(file_path=file_path, original_filename=filename, size_bytes=size_bytes)
        movie.files.append(mfile)
        return movie.id

    def _register_tv_sequential(self, lib: Library, file_path: str, filename: str, parsed, size_bytes: int, cache: dict):
        clean_title = parsed.title.strip()
        lower_title = clean_title.lower()
        
        if lower_title in cache:
            show = cache[lower_title]
        else:
            show = TVShow(library_id=lib.id, title=clean_title, status="unmatched")
            self.db.add(show)
            self.db.flush()
            cache[lower_title] = show

        season_num = parsed.season or 1
        season_key = f"{show.id}_{season_num}"
        if not hasattr(self, '_season_cache'): self._season_cache = {}
        
        season = self._season_cache.get(season_key)
        if not season:
            season = self.db.query(Season).filter(Season.show_id == show.id, Season.season_number == season_num).first()
            if not season:
                season = Season(season_number=season_num)
                show.seasons.append(season)
                self.db.flush() # Ensure season gets ID for episodes
            self._season_cache[season_key] = season

        ep = Episode(
            episode_number=parsed.episode or 1,
            title=f"Episode {parsed.episode or 1}",
            file_path=file_path,
            original_filename=filename
        )
        season.episodes.append(ep)
        return show.id

File: internal/services/scanner.go
package services

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"sync/atomic"

	"media-orchestrator/internal/core"
	"media-orchestrator/internal/database"
	"media-orchestrator/internal/models"
)

var SupportedExts = map[string]bool{
	".mkv": true, ".mp4": true, ".avi": true, ".mov": true,
	".webm": true, ".flv": true, ".iso": true, ".m2ts": true,
	".ts": true,
}

func ScanLibrary(libraryID uint, taskID string, progressChan chan<- int) {
	var lib models.Library
	if err := database.DB.First(&lib, libraryID).Error; err != nil {
		if taskID != "" {
			core.GlobalTaskManager.UpdateTask(taskID, 0, "error", "Library not found")
		}
		return
	}

	normPath := filepath.ToSlash(lib.Path)
	if _, err := os.Stat(normPath); os.IsNotExist(err) {
		if taskID != "" {
			core.GlobalTaskManager.UpdateTask(taskID, 0, "error", "Path does not exist")
		}
		return
	}

	if taskID != "" {
		core.GlobalTaskManager.UpdateTask(taskID, 5, "running", "Discovering files...")
	}

	// 1. Build Caches
	existingPaths := make(map[string]bool)
	if lib.Type == "tv" {
		var episodes []models.Episode
		database.DB.Joins("JOIN seasons ON seasons.id = episodes.season_id").
			Joins("JOIN tv_shows ON tv_shows.id = seasons.show_id").
			Where("tv_shows.library_id = ?", libraryID).
			Select("episodes.file_path").Find(&episodes)
		for _, ep := range episodes {
			existingPaths[ep.FilePath] = true
		}
	} else {
		var mfiles []models.MovieFile
		database.DB.Joins("JOIN movies ON movies.id = movie_files.movie_id").
			Where("movies.library_id = ?", libraryID).
			Select("movie_files.file_path").Find(&mfiles)
		for _, mf := range mfiles {
			existingPaths[mf.FilePath] = true
		}
	}

	var filesToProcess []string
	skipDirs := map[string]bool{
		".git": true, "node_modules": true, "@eaDir": true, "#recycle": true, ".actors": true,
	}

	filepath.WalkDir(normPath, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			name := d.Name()
			if strings.HasPrefix(name, ".") || skipDirs[name] {
				return filepath.SkipDir
			}
			return nil
		}
		ext := strings.ToLower(filepath.Ext(path))
		if SupportedExts[ext] {
			fPath := filepath.ToSlash(path)
			if !existingPaths[fPath] {
				filesToProcess = append(filesToProcess, fPath)
			}
		}
		return nil
	})

	total := len(filesToProcess)
	if total == 0 {
		if taskID != "" {
			core.GlobalTaskManager.UpdateTask(taskID, 100, "done", "No new media files found")
		}
		return
	}

	// 2. Parallel Processing with Worker Pool
	numWorkers := 4
	jobs := make(chan string, total)
	var processedCount int32
	var wg sync.WaitGroup

	// Mutexes for cache-like lookups during registration to avoid DB race conditions on FirstOrCreate patterns
	var registrationMu sync.Mutex

	worker := func() {
		defer wg.Done()
		for fPath := range jobs {
			filename := filepath.Base(fPath)
			parsed := ParseFilename(filename)

			registrationMu.Lock()
			if lib.Type == "tv" {
				processTVFile(lib, fPath, filename, parsed)
			} else {
				processMovieFile(lib, fPath, filename, parsed)
			}
			registrationMu.Unlock()

			newCount := atomic.AddInt32(&processedCount, 1)
			if newCount%10 == 0 || int(newCount) == total {
				progress := 10 + int((float64(newCount)/float64(total))*90)
				if progressChan != nil {
					progressChan <- progress
				}
				if taskID != "" {
					core.GlobalTaskManager.UpdateTask(taskID, progress, "running", fmt.Sprintf("Processing %d/%d: %s", newCount, total, filename))
				}
			}
		}
	}

	for w := 1; w <= numWorkers; w++ {
		wg.Add(1)
		go worker()
	}

	for _, fPath := range filesToProcess {
		jobs <- fPath
	}
	close(jobs)
	wg.Wait()

	if taskID != "" {
		core.GlobalTaskManager.UpdateTask(taskID, 100, "done", fmt.Sprintf("Scan complete. Processed %d new items.", total))
	}
}

func processTVFile(lib models.Library, fPath string, filename string, parsed ParsedMedia) {
	meta := ParseTVShowNFO(fPath)
	title := parsed.Title
	if meta != nil && meta.Title != "" {
		title = meta.Title
	}

	var show models.TVShow
	database.DB.Where("LOWER(title) = ? AND library_id = ?", strings.ToLower(title), lib.ID).First(&show)
	if show.ID == 0 {
		show = models.TVShow{LibraryID: lib.ID, Title: title, Year: parsed.Year}
		database.DB.Create(&show)
	}

	if meta != nil && show.Status != "matched" {
		updates := map[string]interface{}{}
		if meta.Plot != "" { updates["plot"] = meta.Plot }
		if meta.Year != 0 { updates["year"] = meta.Year }
		if meta.TmdbID != "" { updates["tmdb_id"] = meta.TmdbID }
		if meta.ImdbID != "" { updates["imdb_id"] = meta.ImdbID }
		if meta.PosterPath != "" { updates["poster_path"] = meta.PosterPath }
		if meta.FanartPath != "" { updates["fanart_path"] = meta.FanartPath }
		if len(meta.Cast) > 0 { updates["cast"] = meta.Cast }
		if len(updates) > 0 {
			updates["status"] = "matched"
			database.DB.Model(&show).Updates(updates)
		}
	}

	seasonNum := parsed.Season
	if seasonNum == 0 { seasonNum = 1 }
	var season models.Season
	database.DB.Where(models.Season{ShowID: show.ID, SeasonNumber: seasonNum}).FirstOrCreate(&season)

	epNum := parsed.Episode
	if epNum == 0 { epNum = 1 }
	var ep models.Episode
	if err := database.DB.Where("file_path = ?", fPath).First(&ep).Error; err != nil {
		ep = models.Episode{
			SeasonID:         season.ID,
			EpisodeNumber:    epNum,
			Title:            fmt.Sprintf("Episode %d", epNum),
			FilePath:         fPath,
			OriginalFilename: filename,
		}
		database.DB.Create(&ep)
	}
}

func processMovieFile(lib models.Library, fPath string, filename string, parsed ParsedMedia) {
	meta := ParseMovieNFO(fPath)
	title := parsed.Title
	year := parsed.Year
	if meta != nil && meta.Title != "" {
		title = meta.Title
		if meta.Year != 0 {
			year = meta.Year
		}
	}

	var movie models.Movie
	query := database.DB.Where("LOWER(title) = ? AND library_id = ?", strings.ToLower(title), lib.ID)
	if year > 0 {
		query = query.Where("year = ?", year)
	}
	query.First(&movie)

	if movie.ID == 0 && year > 0 {
		database.DB.Where("LOWER(title) = ? AND library_id = ? AND (year = 0 OR year IS NULL)", strings.ToLower(title), lib.ID).First(&movie)
		if movie.ID != 0 {
			movie.Year = year
			database.DB.Model(&movie).Update("year", year)
		}
	}

	if movie.ID == 0 {
		movie = models.Movie{LibraryID: lib.ID, Title: title, Year: year}
		database.DB.Create(&movie)
	}

	if meta != nil && movie.Status != "matched" {
		updates := map[string]interface{}{}
		if meta.Plot != "" { updates["plot"] = meta.Plot }
		if meta.Year != 0 { updates["year"] = meta.Year }
		if meta.TmdbID != "" { updates["tmdb_id"] = meta.TmdbID }
		if meta.ImdbID != "" { updates["imdb_id"] = meta.ImdbID }
		if meta.PosterPath != "" { updates["poster_path"] = meta.PosterPath }
		if meta.FanartPath != "" { updates["fanart_path"] = meta.FanartPath }
		if meta.Runtime != 0 { updates["runtime"] = meta.Runtime }
		if len(meta.Cast) > 0 { updates["cast"] = meta.Cast }
		if len(updates) > 0 {
			updates["status"] = "matched"
			database.DB.Model(&movie).Updates(updates)
		}
	}

	var mfile models.MovieFile
	if err := database.DB.Where("file_path = ?", fPath).First(&mfile).Error; err != nil {
		mfile = models.MovieFile{
			MovieID:          movie.ID,
			FilePath:         fPath,
			OriginalFilename: filename,
		}
		database.DB.Create(&mfile)
	}
}

